### PR TITLE
User Preference of Factors and Readjustment of Weights in the Utility Function

### PR DESCRIPTION
**Pull request by SemanBoon**
## What type of PR is this?

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR introduces a new feature to the booking page that allows users to specify their priority factor when booking a service. The available priority factors are:

- Focus Block (appointments immediately before or after another appointment)
- User Preferred Time (based on past booking trends)
- Popular Slots (popular times for the service provider)

 Based on the selected priority, the weights in the utility function of the recommendation algorithm are adjusted to provide more relevant time slot recommendations.

## Screenshots, Recordings

Screenshot 2024-07-31 at 11 26 44 AM

Screenshot 2024-07-31 at 11 27 04 AM

Screenshot 2024-07-31 at 11 27 11 AM
Screenshot 2024-07-31 at 11 27 19 AM

## Test Plan

Manual Testing:
- Go to the booking page.
- Pick a service from the dropdown menu.
- Select a priority factor from the new dropdown.
- Made sure that the recommended slots update based on what I selected.

Unit Tests: Added unit tests to verify that the utility function adjusts weights correctly based on the selected priority.

### CODE REVIEW

File: Backend/utils.js [hunk 1]
```diff
@@ -54,10 +54,16 @@ const getAvailableSlots = (schedule, serviceDuration, slotPopularity) => {
             }
         }
     }
+    if (availableSlots.length === 0) {
```
Comment (karen-ka) [inline Backend/utils.js:57]: Do we need this? Since we already define availableSlots as [] on Line 39

File: Backend/utils.js [hunk 2]
```diff
@@ -54,10 +54,16 @@ const getAvailableSlots = (schedule, serviceDuration, slotPopularity) => {
...
+        return [];
+    }
     return availableSlots;
 };
 
 const getRecommendedSlots = (slots, userPreferences, provider, currentTime, serviceDuration, preferredPeriod, weights) => {
+    if (slots.length === 0) {
```
Comment (karen-ka) [inline Backend/utils.js:64]: nice!

File: Backend/utils.js [hunk 3]
```diff
@@ -75,11 +81,15 @@ const calculateUserPreferences = (bookings) => {
...
-    const sortedTimes = Object.keys(timeCount).sort((a, b) => timeCount[b] - timeCount[a]);
-
-    // Assume the top 3 most booked times are preferred times
+    const sortedTimes = Object.keys(timeCount).sort((a, b) => {
+        const frequencyDiff = timeCount[b] - timeCount[a];
+        if (frequencyDiff !== 0) {
+            return frequencyDiff;
+        }
+        // If frequencies are equal, sort by time in ascending order
+        return new Date(`1970-01-01T${a}:00Z`) - new Date(`1970-01-01T${b}:00Z`);
```
Comment (karen-ka) [inline Backend/utils.js:91]: Nice tiebreaker

**Review by karen-ka — APPROVED** [review]

File: Backend/utils.js (hunk 1 above)
Comment (SemanBoon) [reply Backend/utils.js:57]: I believe they are checking for different things. Line 39 checks if the service duration is 0, line 60 checks if there were no available slots found. There could be a case where the service duration is not 0 but the available slots could still be empty.

**Pull request merged**